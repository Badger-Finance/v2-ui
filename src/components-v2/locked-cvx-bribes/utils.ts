import BigNumber from 'bignumber.js';
import { formatWithoutExtraZeros, numberWithCommas, unscale } from '../../mobx/utils/helpers';

export function formatBalance(balance: BigNumber): string {
  return numberWithCommas(formatWithoutExtraZeros(unscale(balance, 18)));
}
