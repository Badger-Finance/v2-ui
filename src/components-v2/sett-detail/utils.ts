import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/setts/sett-charts';
import { DelaySeverity } from '../../mobx/model/setts/sett-rewards';
import { SettDetailMode } from '../../mobx/model/setts/sett-detail';

dayjs.extend(utc);

export const ChartModeTitles = {
	[ChartMode.value]: 'Sett Value',
	[ChartMode.ratio]: 'Token Ratio',
	[ChartMode.accountBalance]: 'Your Holdings',
};

export const SettModeTitles = {
	[SettDetailMode.settInformation]: 'Vault Info',
	[SettDetailMode.userInformation]: 'User Info',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};
