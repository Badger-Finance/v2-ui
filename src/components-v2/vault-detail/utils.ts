import { Vault, VaultData } from '@badger-dao/sdk';
import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { ChartMode } from '../../mobx/model/vaults/vault-charts';
import { DelaySeverity } from '../../mobx/model/vaults/vault-rewards';

dayjs.extend(utc);

export const ChartModeTitles: Record<string, string> = {
	[ChartMode.Value]: 'Sett Value',
	[ChartMode.Ratio]: 'Token Ratio',
	[ChartMode.AccountBalance]: 'My Holdings',
	[ChartMode.BoostMultiplier]: 'Badger Boost',
};

export const calculateDelaySeverity = (delay: number): DelaySeverity => {
	if (delay >= 4) {
		return DelaySeverity.high;
	}

	if (delay >= 2) {
		return DelaySeverity.medium;
	}

	return DelaySeverity.none;
};

export const calculateDifferenceInHoursFromCycle = (cycle: Date): number => {
	return Math.abs(dayjs(cycle).diff(dayjs(), 'hours'));
};

export function defaultVaultBalance(vault: Vault): VaultData {
	return {
		address: vault.vaultToken,
		name: vault.name,
		symbol: vault.asset,
		pricePerFullShare: vault.pricePerFullShare,
		balance: 0,
		value: 0,
		earnedBalance: 0,
		earnedValue: 0,
		depositedBalance: 0,
		withdrawnBalance: 0,
		tokens: [],
		earnedTokens: [],
	};
}

export interface Balance {
	balance: number;
}

export function hasBalance(input?: Balance): boolean {
	if (!input) {
		return false;
	}
	return input.balance > 0;
}
