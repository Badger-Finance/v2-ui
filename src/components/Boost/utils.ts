import { makeStyles } from '@material-ui/core/styles';
import { ClassNameMap } from '@material-ui/core/styles/withStyles';
import React from 'react';

type ComparisonConfig = {
  toCompareValue: number;
  toBeComparedValue: number;
  greaterCaseColor: string;
  lessCaseColor: string;
  defaultColor: string;
};

export const getColorFromComparison = ({
  toCompareValue,
  toBeComparedValue,
  greaterCaseColor,
  lessCaseColor,
  defaultColor,
}: ComparisonConfig): string => {
  if (toCompareValue > toBeComparedValue) return greaterCaseColor;
  if (toCompareValue < toBeComparedValue) return lessCaseColor;

  return defaultColor;
};

export const useAssetInputStyles = (
  currentValue: string,
  toCompare = 0,
): ((props?: React.CSSProperties) => ClassNameMap<'assetColor'>) => {
  return makeStyles((theme) => {
    const defaultColor = currentValue ? theme.palette.text.primary : theme.palette.text.secondary;
    const fontColor = getColorFromComparison({
      toCompareValue: Number(currentValue),
      toBeComparedValue: toCompare,
      greaterCaseColor: '#74D189',
      lessCaseColor: theme.palette.error.main,
      defaultColor,
    });

    return {
      assetColor: {
        color: fontColor,
      },
    };
  });
};
