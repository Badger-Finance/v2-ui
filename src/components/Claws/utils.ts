import BigNumber from 'bignumber.js';
import { INVALID_REASON } from './index';

export interface ValidationParams {
	amount: string | BigNumber;
	maximum: BigNumber;
	minimum?: BigNumber;
}

export function validateAmountBoundaries({
	amount,
	maximum,
	minimum = new BigNumber('0'),
}: ValidationParams): INVALID_REASON | undefined {
	let error: INVALID_REASON | undefined = undefined;

	const input = new BigNumber(amount);
	const amountExceedsBalance = input.gt(maximum);
	const amountIsLessThanMinimum = input.lt(minimum);

	if (amountIsLessThanMinimum) error = INVALID_REASON.UNDER_MINIMUM;
	if (amountExceedsBalance) error = INVALID_REASON.OVER_MAXIMUM;

	console.log({
		input: input.toString(),
		minimum: minimum.toString(),
		maximum: maximum.toString(),
		amountExceedsBalance,
		amountIsLessThanMinimum,
		error,
	});

	return error;
}
