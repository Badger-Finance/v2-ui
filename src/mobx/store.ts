import { RouterStore } from 'mobx-router';
import UiState from './reducers';
import WalletStore from './stores/walletStore';
import ContractsStore from './stores/contractsStore';
import AirdropStore from './stores/airdropStore';
import RebaseStore from './stores/rebaseStore';
import RewardsStore from './stores/rewardsStore';
import IbBTCStore from './stores/ibBTCStore';
import TransactionsStore from './stores/transactionsStore';
import SettStoreV2 from './stores/settStoreV2';
import { NETWORK_LIST } from '../config/constants';
import HoneyPotStore from './stores/honeyPotStore';

export class RootStore {
	public router: RouterStore<RootStore>;
	public wallet: WalletStore;
	public uiState: UiState;
	public contracts: ContractsStore;
	public airdrops: AirdropStore;
	public rebase: RebaseStore;
	public rewards: RewardsStore;
	public ibBTCStore: IbBTCStore;
	public setts: SettStoreV2;
	public transactions: TransactionsStore;
	public honeyPot: HoneyPotStore;

	constructor() {
		this.router = new RouterStore<RootStore>(this);
		this.wallet = new WalletStore(this);
		this.contracts = new ContractsStore(this);
		this.airdrops = new AirdropStore(this);
		this.rebase = new RebaseStore(this);
		this.rewards = new RewardsStore(this);
		this.uiState = new UiState(this);
		this.ibBTCStore = new IbBTCStore(this);
		this.honeyPot = new HoneyPotStore(this);
		// RenVM transactions store.
		this.transactions = new TransactionsStore(this);
		this.setts = new SettStoreV2(this);

		this.walletRefresh();
	}

	async walletRefresh(): Promise<void> {
		const chain = this.wallet.network.name;
		const refreshData = [
			this.setts.loadAssets(chain),
			this.setts.loadPrices(chain),
			this.wallet.getGasPrice(),
			this.contracts.updateProvider(),
		];
		if (chain === NETWORK_LIST.ETH) {
			refreshData.push(this.setts.loadGeysers(chain));
			refreshData.push(this.rebase.fetchRebaseStats());
		} else {
			refreshData.push(this.setts.loadSetts(chain));
		}
		await Promise.all(refreshData);

		if (this.wallet.connectedAddress) {
			this.contracts.updateProvider();
			await this.wallet.getGasPrice();
			await this.contracts.fetchContracts();
			if (chain === NETWORK_LIST.ETH) {
				this.uiState.reduceRebase();
				this.ibBTCStore.init();
				this.rewards.fetchSettRewards();
				this.uiState.reduceTreeRewards();
				this.airdrops.fetchAirdrops();
				this.uiState.reduceAirdrops();
			}
			this.uiState.reduceStats();
		}
	}
}

const store = new RootStore();

export default store;
