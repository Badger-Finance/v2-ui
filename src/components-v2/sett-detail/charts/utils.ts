import dayjs from 'dayjs';
import utc from 'dayjs/plugin/utc';
import { Sett } from '../../../mobx/model/setts/sett';
import { fetchSettChartInformation } from '../../../mobx/utils/apiV2';
import { SettSnapshotGranularity } from '../../../mobx/model/setts/sett-snapshot';
import { SettChartData } from '../../../mobx/model/setts/sett-charts';

dayjs.extend(utc);

export enum SettChartTimeframe {
	'day' = 1,
	'week' = 7,
	'month' = 30,
}

/**
 * Fetches chart information since the provided timeframe until the current date
 * @param sett
 * @param timeframe
 */
export const fetchSettChart = async (sett: Sett, timeframe: SettChartTimeframe): Promise<SettChartData[] | null> => {
	const timeframeDays = 1 * timeframe;
	const isDayTimeFrame = timeframe === SettChartTimeframe.day;

	// if timeframe is just one day then we want the granularity to be hours
	const granularity = isDayTimeFrame ? SettSnapshotGranularity.HOUR : SettSnapshotGranularity.DAY;

	const now = dayjs().utc(); // query until current date
	const from = dayjs(now).subtract(timeframeDays, 'days').utc();

	const fetchedData = await fetchSettChartInformation(sett.vaultToken, from.toDate(), now.toDate(), granularity);

	if (!fetchedData) {
		return null;
	}

	// data needs to be ascending sorted
	return fetchedData
		.sort((a, b) => a.timestamp - b.timestamp)
		.map((d) => ({ ...d, timestamp: new Date(d.timestamp) }));
};
