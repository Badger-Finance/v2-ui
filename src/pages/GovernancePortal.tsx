import AddressInfoCard from '../components-v2/governance-portal/AddressInfoCard';
import EventsTable from '../components-v2/governance-portal/EventsTable';
import { Grid } from '@material-ui/core';
import PageHeader from '../components-v2/common/PageHeader';
import { observer } from 'mobx-react-lite';
import React, { useEffect, useContext } from 'react';
import { HeaderContainer, LayoutContainer } from '../components-v2/common/Containers';
import { StoreContext } from '../mobx/store-context';

const GovernancePortal = observer(() => {
	const store = useContext(StoreContext);
	const { governancePortal } = store;

	useEffect(() => {
		governancePortal.loadData();
	});

	return (
		<LayoutContainer>
			<Grid container spacing={1} justify="center">
				<HeaderContainer item xs={12}>
					<PageHeader title="Governance Portal" subtitle="Review recent activity from the DAO." />
				</HeaderContainer>

				<Grid item xs={12} md={4}>
					<AddressInfoCard title="Timelock Contract" address={governancePortal.contractAddress} />
				</Grid>

				<Grid item xs={12} md={4}>
					<AddressInfoCard title="Admin" address={governancePortal.adminAddress} />
				</Grid>

				<Grid item xs={12} md={4}>
					<AddressInfoCard title="Guardian" address={governancePortal.guardianAddress} />
				</Grid>
			</Grid>
			<Grid container justify="center">
				<Grid item xs={12} md={9}>
					<EventsTable events={governancePortal.timelockEvents} />
				</Grid>
			</Grid>
		</LayoutContainer>
	);
});

export default GovernancePortal;
