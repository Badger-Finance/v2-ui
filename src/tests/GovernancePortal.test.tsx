import React from 'react';
import store from '../mobx/RootStore';
import GovernancePortal from '../../src/pages/GovernancePortal';
import { checkSnapshot } from './utils/snapshots';
import { action } from 'mobx';

describe('Governance Portal', () => {
	beforeEach(() => {
		store.governancePortal.loadData = action(jest.fn());
		store.governancePortal.adminAddress = '0xB65cef03b9B89f99517643226d76e286ee999e77';
		store.governancePortal.contractAddress = '0x21CF9b77F88Adf8F8C98d7E33Fe601DC57bC0893';
		store.governancePortal.guardianAddress = '0x576cD258835C529B54722F84Bb7d4170aA932C64';
		store.governancePortal.timelockEvents = [
			{
				blockNumber: 12270941,
				returnValues: {
					signature: 'upgrade(address,address)',
					data:
						'0x0000000000000000000000006def55d2e18486b9ddfaa075bc4e4ee0b28c15450000000000000000000000005c7adb3fd0df2d1822a36922dd941e16d2bf4e51',
				},
				event: 'QueueTransaction',
				functionName: 'upgrade',
				parameterTypes: ['address', 'address'],
				decodedParameters: {
					'0': '0x6dEf55d2e18486B9dDfaA075bc4e4EE0B28c1545',
					'1': '0x5c7AdB3Fd0DF2D1822a36922dd941e16D2bF4E51',
				},
			},
		];
	});

	it('renders correctly', () => {
		checkSnapshot(<GovernancePortal />);
	});
});
