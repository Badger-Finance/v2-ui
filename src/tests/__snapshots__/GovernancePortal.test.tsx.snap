// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Governance Portal renders correctly 1`] = `
<div>
  <div
    class="MuiContainer-root-2 WithStyles(ForwardRef(Container))-root-1 MuiContainer-maxWidthLg-8"
  >
    <div
      class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-spacing-xs-1-33 MuiGrid-justify-xs-center-28"
    >
      <div
        class="MuiGrid-root-10 WithStyles(ForwardRef(Grid))-root-113 WithStyles(ForwardRef(Grid))-root-114 MuiGrid-item-12 MuiGrid-grid-xs-12-56"
      >
        <div
          class="makeStyles-headerContainer-116"
        >
          <div
            class="makeStyles-headerContent-117"
          >
            <h2
              class="MuiTypography-root-118 MuiTypography-h2-124 MuiTypography-colorTextPrimary-143"
            >
              Governance Portal
            </h2>
            <h6
              class="MuiTypography-root-118 MuiTypography-subtitle1-129 MuiTypography-colorTextSecondary-144"
            >
              Review recent activity from the DAO.
            </h6>
          </div>
          <div
            class="makeStyles-headerWidgets-115"
          >
            <div
              class="MuiInputBase-root-177 MuiOutlinedInput-root-162 MuiInputBase-colorSecondary-186 MuiOutlinedInput-colorSecondary-163 makeStyles-gasSelector-148 MuiInputBase-adornedStart-181 MuiOutlinedInput-adornedStart-166"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-195 makeStyles-gasIcon-149"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M19.77 7.23l.01-.01-3.72-3.72L15 4.56l2.11 2.11c-.94.36-1.61 1.26-1.61 2.33 0 1.38 1.12 2.5 2.5 2.5.36 0 .69-.08 1-.21v7.21c0 .55-.45 1-1 1s-1-.45-1-1V14c0-1.1-.9-2-2-2h-1V5c0-1.1-.9-2-2-2H6c-1.1 0-2 .9-2 2v16h10v-7.5h1.5v5c0 1.38 1.12 2.5 2.5 2.5s2.5-1.12 2.5-2.5V9c0-.69-.28-1.32-.73-1.77zM12 10H6V5h6v5zm6 0c-.55 0-1-.45-1-1s.45-1 1-1 1 .45 1 1-.45 1-1 1z"
                />
              </svg>
              <div
                aria-haspopup="listbox"
                class="MuiSelect-root-151 MuiSelect-select-152 MuiSelect-selectMenu-155 MuiSelect-outlined-154 MuiInputBase-input-188 MuiOutlinedInput-input-172 MuiInputBase-inputAdornedStart-192 MuiOutlinedInput-inputAdornedStart-175"
                role="button"
                tabindex="0"
              >
                25
              </div>
              <input
                aria-hidden="true"
                class="MuiSelect-nativeInput-161"
                tabindex="-1"
                value="standard"
              />
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root-195 MuiSelect-icon-157 MuiSelect-iconOutlined-160"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7 10l5 5 5-5z"
                />
              </svg>
              <fieldset
                aria-hidden="true"
                class="PrivateNotchedOutline-root-208 MuiOutlinedInput-notchedOutline-171"
                style="padding-left: 8px;"
              >
                <legend
                  class="PrivateNotchedOutline-legend-209"
                  style="width: 0.01px;"
                >
                  <span>
                    â€‹
                  </span>
                </legend>
              </fieldset>
            </div>
            <button
              class="MuiButtonBase-root-245 MuiButton-root-216 MuiButton-contained-224 makeStyles-walletButton-215 MuiButton-containedSecondary-226 MuiButton-disableElevation-227"
              tabindex="0"
              type="button"
            >
              <span
                class="MuiButton-label-217"
              >
                Click to connect
                <span
                  class="MuiButton-endIcon-241 MuiButton-iconSizeMedium-243"
                >
                  <div
                    class="makeStyles-walletDot-212 makeStyles-redDot-213"
                  />
                </span>
              </span>
              <span
                class="MuiTouchRipple-root-323"
              />
            </button>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-md-4-76"
      >
        <div
          class="MuiPaper-root-250 makeStyles-infoPaper-248 MuiPaper-elevation2-255 MuiPaper-rounded-251"
        >
          <h6
            class="MuiTypography-root-118 MuiTypography-subtitle1-129 MuiTypography-colorTextPrimary-143"
          >
            Timelock Contract
          </h6>
          <div
            class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-justify-xs-center-28"
          >
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-9-53"
            >
              <div
                class="MuiBox-root-278 MuiBox-root-279"
              >
                <h5
                  class="MuiTypography-root-118 MuiTypography-h5-127 MuiTypography-noWrap-137"
                >
                  0x21CF9b77F88Adf8F8C98d7E33Fe601DC57bC0893
                </h5>
              </div>
            </div>
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-auto-43"
            >
              <a
                class="MuiTypography-root-118 MuiLink-root-280 MuiLink-underlineHover-282 makeStyles-linkIcon-249 MuiTypography-colorPrimary-141"
                href="https://etherscan.io/address/0x21CF9b77F88Adf8F8C98d7E33Fe601DC57bC0893"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-195"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-md-4-76"
      >
        <div
          class="MuiPaper-root-250 makeStyles-infoPaper-248 MuiPaper-elevation2-255 MuiPaper-rounded-251"
        >
          <h6
            class="MuiTypography-root-118 MuiTypography-subtitle1-129 MuiTypography-colorTextPrimary-143"
          >
            Admin
          </h6>
          <div
            class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-justify-xs-center-28"
          >
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-9-53"
            >
              <div
                class="MuiBox-root-278 MuiBox-root-286"
              >
                <h5
                  class="MuiTypography-root-118 MuiTypography-h5-127 MuiTypography-noWrap-137"
                >
                  0xB65cef03b9B89f99517643226d76e286ee999e77
                </h5>
              </div>
            </div>
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-auto-43"
            >
              <a
                class="MuiTypography-root-118 MuiLink-root-280 MuiLink-underlineHover-282 makeStyles-linkIcon-249 MuiTypography-colorPrimary-141"
                href="https://etherscan.io/address/0xB65cef03b9B89f99517643226d76e286ee999e77"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-195"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
      <div
        class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-md-4-76"
      >
        <div
          class="MuiPaper-root-250 makeStyles-infoPaper-248 MuiPaper-elevation2-255 MuiPaper-rounded-251"
        >
          <h6
            class="MuiTypography-root-118 MuiTypography-subtitle1-129 MuiTypography-colorTextPrimary-143"
          >
            Guardian
          </h6>
          <div
            class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-justify-xs-center-28"
          >
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-9-53"
            >
              <div
                class="MuiBox-root-278 MuiBox-root-287"
              >
                <h5
                  class="MuiTypography-root-118 MuiTypography-h5-127 MuiTypography-noWrap-137"
                >
                  0x576cD258835C529B54722F84Bb7d4170aA932C64
                </h5>
              </div>
            </div>
            <div
              class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-auto-43"
            >
              <a
                class="MuiTypography-root-118 MuiLink-root-280 MuiLink-underlineHover-282 makeStyles-linkIcon-249 MuiTypography-colorPrimary-141"
                href="https://etherscan.io/address/0x576cD258835C529B54722F84Bb7d4170aA932C64"
                target="_blank"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root-195"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M10.09 15.59L11.5 17l5-5-5-5-1.41 1.41L12.67 11H3v2h9.67l-2.58 2.59zM19 3H5c-1.11 0-2 .9-2 2v4h2V5h14v14H5v-4H3v4c0 1.1.89 2 2 2h14c1.1 0 2-.9 2-2V5c0-1.1-.9-2-2-2z"
                  />
                </svg>
              </a>
            </div>
          </div>
        </div>
      </div>
    </div>
    <div
      class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-justify-xs-center-28"
    >
      <div
        class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-md-9-81"
      >
        <h4
          class="MuiTypography-root-118 makeStyles-title-289 MuiTypography-h4-126"
        >
          Recent Timelock Activity
        </h4>
        <div
          class="MuiGrid-root-10 makeStyles-root-288 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-grid-xs-12-56"
        >
          <div
            class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-sm-3-61"
          >
            <p
              class="MuiTypography-root-118 MuiTypography-body2-119 MuiTypography-colorTextSecondary-144"
            >
              Block Number
            </p>
          </div>
          <div
            class="MuiGrid-root-10 makeStyles-hiddenMobile-290 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-sm-true-58"
          >
            <p
              class="MuiTypography-root-118 MuiTypography-body2-119 MuiTypography-colorTextSecondary-144"
            >
              Event
            </p>
          </div>
          <div
            class="MuiGrid-root-10 makeStyles-hiddenMobile-290 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-sm-true-58"
          >
            <p
              class="MuiTypography-root-118 MuiTypography-body2-119 MuiTypography-colorTextSecondary-144"
            >
              Action
            </p>
          </div>
        </div>
        <ul
          class="MuiList-root-292 makeStyles-list-291 MuiList-padding-293"
        >
          <li
            class="MuiListItem-root-300 makeStyles-listItem-297 MuiListItem-gutters-307"
          >
            <div
              class="MuiGrid-root-10 makeStyles-root-296 MuiGrid-container-11"
            >
              <div
                class="MuiGrid-root-10 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-grid-xs-12-56"
              >
                <div
                  class="MuiGrid-root-10 makeStyles-name-299 MuiGrid-container-11 MuiGrid-item-12 MuiGrid-grid-xs-12-56 MuiGrid-grid-sm-3-61"
                >
                  12270941
                </div>
                <div
                  class="MuiGrid-root-10 makeStyles-mobileLabel-298 MuiGrid-item-12 MuiGrid-grid-xs-6-50 MuiGrid-grid-md-true-72"
                >
                  <p
                    class="MuiTypography-root-118 MuiTypography-body2-119 MuiTypography-colorTextSecondary-144"
                  >
                    Event
                  </p>
                </div>
                <div
                  class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-6-50 MuiGrid-grid-md-true-72"
                >
                  QueueTransaction
                </div>
                <div
                  class="MuiGrid-root-10 makeStyles-mobileLabel-298 MuiGrid-item-12 MuiGrid-grid-xs-6-50 MuiGrid-grid-md-true-72"
                >
                  <p
                    class="MuiTypography-root-118 MuiTypography-body2-119 MuiTypography-colorTextSecondary-144"
                  >
                    Action
                  </p>
                </div>
                <div
                  class="MuiGrid-root-10 MuiGrid-item-12 MuiGrid-grid-xs-6-50 MuiGrid-grid-md-true-72"
                >
                  upgrade
                  <span>
                    (
                  </span>
                  <span>
                    <span
                      class="makeStyles-tooltipWrap-311"
                      title="0x6dEf55d2e18486B9dDfaA075bc4e4EE0B28c1545"
                    >
                      address
                    </span>
                    , 
                  </span>
                  <span>
                    <span
                      class="makeStyles-tooltipWrap-311"
                      title="0x5c7AdB3Fd0DF2D1822a36922dd941e16D2bF4E51"
                    >
                      address
                    </span>
                  </span>
                  <span>
                    )
                  </span>
                </div>
              </div>
            </div>
          </li>
        </ul>
      </div>
    </div>
  </div>
</div>
`;
